File: 130.surrounded-regions.js
/*
 * @lc app=leetcode id=130 lang=javascript
 *
 * [130] Surrounded Regions
 */

// @lc code=start
/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var solve = function (board) {
  const m = board.length;
  const n = board[0].length;
  for (let i = 0; i < m; ++i) {
    for (let j = 0; j < n; ++j) {
      if ((0 < i && i < m - 1 && 0 < j && j < n - 1) || board[i][j] !== "O") {
        continue;
      }
      const q = [[i, j]];
      while (q.length) {
        const [x, y] = q.shift();
        if (0 <= x && x < m && 0 <= y && y < n && board[x][y] === "O") {
          board[x][y] = "#";
          q.push([x, y + 1], [x, y - 1], [x + 1, y], [x - 1, y]);
        }
      }
    }
  }
  for (let i = 0; i < m; ++i) {
    for (let j = 0; j < n; ++j) {
      if (board[i][j] === "O") {
        board[i][j] = "X";
      } else if (board[i][j] === "#") {
        board[i][j] = "O";
      }
    }
  }
};
// @lc code=end

File: 130.surrounded-regions.py
#
# @lc app=leetcode id=130 lang=python3
#
# [130] Surrounded Regions
#

# Time:  O(m * n)
# Space: O(m + n)

from collections import deque
from typing import List


# @lc code=start
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        def helper(x, y):
            if 0 <= x < m and 0 <= y < n and board[x][y] == "O":
                board[x][y] = "#"
                helper(x - 1, y), helper(x + 1, y)
                helper(x, y - 1), helper(x, y + 1)

        m, n = len(board), len(board[0])

        # bfs
        for i in range(m):
            for j in range(n):
                if 0 < i < m - 1 and 0 < j < n - 1 or board[i][j] != "O":
                    continue
                queue = deque([(i, j)])
                while queue:
                    x, y = queue.popleft()
                    if 0 <= x < m and 0 <= y < n and board[x][y] == "O":
                        board[x][y] = "#"
                        queue.append((x - 1, y)), queue.append((x + 1, y))
                        queue.append((x, y - 1)), queue.append((x, y + 1))
        for i in range(m):
            for j in range(n):
                if board[i][j] == "O":
                    board[i][j] = "X"
                elif board[i][j] == "#":
                    board[i][j] = "O"
        return

        # dfs
        for i in range(1, m - 1):
            helper(i, 0), helper(i, n - 1)
        for j in range(n):
            helper(0, j), helper(m - 1, j)
        for i in range(m):
            for j in range(n):
                if board[i][j] == "O":
                    board[i][j] = "X"
                elif board[i][j] == "#":
                    board[i][j] = "O"


# @lc code=end
